_posts/2020-02-13-spring-boot-application-events-explained.md

```diff
@@ -0,0 +1,263 @@
+---
+title: Spring Boot Application Events Explained
+categories: [spring-boot]
+date: 2020-02-13 05:00:00 +1100
+modified: 2020-02-13 05:00:00 +1100
+author: nandan
+excerpt: 'Spring Boot allows us to throw and listen to specific application events that we can process as we wish. Events are meant for exchanging information between loosely coupled components.'
+image:
+  auto: 0058-motorway-junction
+---
+
+To "listen" to an event, we can always write the listener as another method within the source of the event, but this will tightly couple the event source to the logic of the listener.
+
+With real events, we are more flexible than with direct method calls. We can dynamically register and deregister listeners to certain events as we wish. We can also have multiple listeners for the same event.
+
+This tutorial gives an overview of how to use publish and listen to custom events and explains Spring Boot's built-in events.
+
+{% include github-project.html url="https://github.com/thombergs/code-examples/tree/master/spring-boot/spring-events" %}
+
+## Why Should I Use Events Instead of Direct Method Calls
+
+Both events and direct method calls fits for different situations. With method call its like making assertion that, no matter what the state of modules are, they need to know this event happened. With events you just say this event occured and which modules get notified is not my concern. It is good to use events when you want to pass on the processing to another thread (example: sending an email on some task completion). Also for test driven development events comes in handy.
+
+## What is an Application Event?
+
+Spring application events allows us to throw and listen to specific application events that we can process as we wish. Events are meant for exchanging information between loosely coupled components. As there is no direct coupling between publishers and subscribers, it enables us to modify subscribers without affecting the publishers and vice-versa.
+
+Let's see how we can create, publish and listen to custom events in a Spring Boot application.
+
+## Creating an `ApplicationEvent`
+
+We can publish application events using the Spring Framework’s event publishing mechanism. Prior to Spring 4.2, events had to extend `ApplicationEvent`. In latest versions, we can just publish any object as an event.
+We are passing `source` object on which the event initially occurred or with which the event is associated (never null) to `super`. We do this for event sourcing. In event sourcing, we capture the state of entity or aggregate as a sequence of state changing events.
```

Comment:
Added an empty line to separate paragraphs.

Moved the explanation of `source` to after the code example. Always explain code AFTER the code example, otherwise the reader will be confused.